More LaTeX examples

ianthomas23 opened the pull request:
This PR adds two new LaTeX examples.

The first one is a `duffing_oscillator.py` which is a form of harmonic motion that exhibits chaotic behaviour for certain parameter combinations. It has LaTeX on `Slider`, `Div`, `Paragraph` and axis labels. The following example image and video files were generated using the branch in PR #11815 which provides mathtext on sliders.

![duffing](https://user-images.githubusercontent.com/580326/144893160-3a86f7a6-6a98-4298-8034-d94e8535c09f.png)

https://user-images.githubusercontent.com/580326/144895660-bf213253-e38a-4a3d-ac47-af62eaaab1eb.mov

The second example is `schrodinger.py` which shows the solution to a simple form of Schrodinger's equation and uses LaTeX on `Label` and `Title` annotations. The image below was generated against the branch used in PR #11841.

![schrodinger](https://user-images.githubusercontent.com/580326/144895799-6f68d60e-bd57-4260-a2f3-21f1158c17a5.png)

bryevdv commented:
Wow these are terrific! A few quick initial comments:

* The first example might look nice with a very light grey (`#fafafa` or `#fefefe`) plot background? At least I think a little more weight to cut up the visual field might benice
* These should definitely go in the galler. Will need a thumbnail for each and an entry in the `gallery.json` config

ianthomas23 commented:
@bryevdv I'll do your suggested improvements first thing tomorrow. I am very happy to receive suggestions for improvements to the appearance and the underlying python scripts as I have exhausted myself doing the maths and getting the appearance this far.

ianthomas23 commented:
Changes:
1. Added `background_fill_color="#f6f6f6"` to first example (any whiter looks too subtle):
![duffing2](https://user-images.githubusercontent.com/580326/145026764-40303acb-680d-4e21-bef3-7deee6c2d939.png)
2. Dashed line for potential in second example: 
![schrodinger2](https://user-images.githubusercontent.com/580326/145026874-fd874973-046a-450f-9f27-ce06cf9d8c7f.png)
3. Added second example to gallery, and built docs locally to check this is OK.

I have two questions/problems now:
1. I put both examples in the directory `examples/plotting/file` but `examples/plotting/README.md` states
```
There are three subdirectories:

* `file` examples that output to static HTML files

* `notebook` examples that display inline in [Jupyter](http://jupyter.org) notebooks

* `server` examples that send output to a Bokeh server, using the python client library for the server
```
so I should put the streaming example in the `server` directory. But there are no `server` or `notebook` directories here. So where should it go? And should I update the README.md whilst I am here?

2. The sphinx output doesn't set up links for `Label` and `Title` as seen here
![screenshot](https://user-images.githubusercontent.com/580326/145027960-2e09e497-885d-49dd-9b53-71376b5e19a0.png)
The same happens for `ColorBar` in the eclipse example, and `LabelSet` in the elements example.

mattpap commented on examples/plotting/file/duffing_oscillator.py:
```diff
@@ -0,0 +1,125 @@
+
+doc = curdoc()
+doc.add_periodic_callback(stream, 20)
+doc.add_root(row(column(col), ps[0], ps[1]))
```
This example requires bokeh server, so it should be moved under `examples/app`.

ianthomas23 replied:
Thanks. Will do.

bryevdv replied:
Just for FYI there are also a few app examples under `examples/howto` but that is because they are specifically tailored as a reference demonstration of some narrow feature of Bokeh server apps in isolation (e.g. UX events, or hold/unhold) and are not otherwise very interesting on their own. The `examples/app` demos generally try to be "interesting" or use real data, etc. I think that is the case here (and we might consider adding this to demo.bokeh.org  as well... cc @tcmetzger we would need to update the gallery thumbnail links if we want to do that)

tcmetzger replied:
@bryevdv that makes sense - and yes, that would be great to have on the gallery page! We'll probably have to wait to actually put it there until 3.0 is published, so people can use the functionalities demonstrated here, so we'll need to remember to do that then.

ianthomas23 commented:
I have moved the server example to `examples/app` and created a new issue to update the example README files (#11868) rather than fix it here.

This is ready AFAIAC.

bryevdv commented on examples/app/duffing_oscillator.py:
```diff
@@ -0,0 +1,125 @@
+    :apis: bokeh.plotting.Figure.circle, bokeh.plotting.Figure.line, bokeh.models.sources.ColumnDataSource.stream,
+        bokeh.models.Div, bokeh.models.Paragraph, bokeh.models.Slider
+    :refs: :ref:`userguide_styling` > :ref:`userguide_styling_math`
+    :keywords: mathtext, latex
+'''
```
nit, but I think all other examples docstrings always end with blank link 
```suggestion
    :keywords: mathtext, latex

'''
```

ianthomas23 replied:
Fixed it.

bryevdv commented:
Thanks @ianthomas23 !
